### PR TITLE
Project 6: Part 1

### PR DESCRIPTION
**Pull request by dpsproservices**
Project 6: Part 1

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -64,14 +64,20 @@ send me a DM to check your pull request
...
-    //3
+
+    int value;
+    std::string name;
+
+    T(int v, const char* s)   //1
+    {
+        value = v; //2
+        name = s; //3
+    }
```
Comment (matkatmusic) [inline main.cpp]: Use the member initializer list to init your class members with the parameters.
The reason: you are initializing them with their default values when they're constructed.
Then you're initializing an instance on the right side of the `=`, which must then invoke the `operator=` to assign them.  That's a double initialization. 
- [ ] revise

File: main.cpp [hunk 2]
```diff
@@ -64,14 +64,20 @@ send me a DM to check your pull request
...
 };
 
-struct <#structName1#>                                //4
+struct X                                //4
 {
-    <#type name#> compare(<#type name#> a, <#type name#> b) //5
+    T* compare(T* a, T* b) //5
     {
         if( a->value < b->value ) return a;
         if( a->value > b->value ) return b;
```
Comment (matkatmusic) [inline main.cpp]: What must we always do before we use a pointer?
- [ ] revise

File: main.cpp [hunk 3]
```diff
@@ -81,53 +87,56 @@ struct <#structName1#>                                //4
...
 {
-    float <#name1#> { 0 }, <#name2#> { 0 };
-    <#returnType#> <#memberFunction#>(<#type name#>* <#updatedValue#>)      //12
+    float x { 0 }, y { 0 };
+    float multiply(float* updatedValue)      //12
     {
-        
+        std::cout << "U's x value: " << this->x << std::endl;
+        this->x = *updatedValue;
+        std::cout << "U's x updated value: " << this->x << std::endl;
```
Comment (matkatmusic) [inline main.cpp]: What must we always do before we use a pointer?
- [ ] revise

File: main.cpp [hunk 4]
```diff
@@ -81,53 +87,56 @@ struct <#structName1#>                                //4
...
-    static <#returntype#> <#staticFunctionA#>(U* that, <#type name#>* <#updatedValue#> )        //10
+    static float multiply(U* that, float* updatedValue )        //10
     {
-        std::cout << "U's <#name1#> value: " << that-><#name1#> << std::endl;
-        that-><#name1#> = <#updatedValue#>;
-        std::cout << "U's <#name1#> updated value: " << that-><#name1#> << std::endl;
-        while( std::abs(that-><#name2#> - that-><#name1#>) > 0.001f )
+        std::cout << "U's x value: " << that->x << std::endl;
+        that->x = *updatedValue;
+        std::cout << "U's x updated value: " << that->x << std::endl;
```
Comment (matkatmusic) [inline main.cpp]: What must we always do before we use a pointer?
- [ ] revise

File: main.cpp [hunk 5]
```diff
@@ -81,53 +87,56 @@ struct <#structName1#>                                //4
...
-    T <#name1#>( , );                                             //6
-    T <#name2#>( , );                                             //6
+    T t1(1 , "David");                                             //6
+    T t2(2 , "Alexander");                                             //6
     
-    <#structName1#> f;                                            //7
-    auto* smaller = f.compare( , );                              //8
+    X f;                                            //7
+    auto* smaller = f.compare(&t1 , &t2);                              //8
     std::cout << "the smaller one is << " << smaller->name << std::endl; //9
```
Comment (matkatmusic) [inline main.cpp]: What must we always do before we use a pointer?
- [ ] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp [hunk 6]
```diff
@@ -109,29 +108,40 @@ struct Z
...
-    std::cout << "the smaller one is << " << smaller->name << std::endl; //9
+    X f; //7
+    auto* smaller = f.compare(&t1 , &t2); //8
+
+    if (smaller != nullptr)
+    {
+        std::cout << "the smaller one is << " << smaller->name << std::endl; //9
+    } else {
+        std::cout << "Either t1 or t2 is not initialized." << std::endl;
+    }
```
Comment (matkatmusic) [inline main.cpp]: What is another reason that `compare` would return a nullptr?  Take a 2nd look at the logic in that function

File: main.cpp [hunk 7]
```diff
@@ -64,70 +64,89 @@ send me a DM to check your pull request
...
 {
-    float <#name1#> { 0 }, <#name2#> { 0 };
-    <#returnType#> <#memberFunction#>(<#type name#>* <#updatedValue#>)      //12
+    float x { 0 }, y { 0 };
+    float multiply(float* updatedValue)      //12
     {
-        
+        std::cout << "U's x value: " << this->x << std::endl;
+        this->x = *updatedValue;
+        std::cout << "U's x updated value: " << this->x << std::endl;
```
Comment (matkatmusic) [inline main.cpp]: You still didn't fix this.

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp (hunk 7 above)
Comment (dpsproservices) [reply main.cpp]: done

File: main.cpp (hunk 6 above)
Comment (dpsproservices) [reply main.cpp]: done

File: main.cpp (hunk 5 above)
Comment (dpsproservices) [reply main.cpp]: check for nullptr

File: main.cpp (hunk 4 above)
Comment (dpsproservices) [reply main.cpp]: check for nullptr

File: main.cpp (hunk 3 above)
Comment (dpsproservices) [reply main.cpp]: check for nullptr

File: main.cpp (hunk 2 above)
Comment (dpsproservices) [reply main.cpp]: check for nullptr

File: main.cpp (hunk 1 above)
Comment (dpsproservices) [reply main.cpp]: done

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**